### PR TITLE
(FACT-2619) External facts cache

### PR DESCRIPTION
**Pull request by florindragos**

### CODE REVIEW

File: lib/custom_facts/util/collection.rb [hunk 1]
```diff
@@ -148,11 +147,12 @@ def to_hash
...
-        fact_value = fact&.value
-        return Facter.core_value(name) if fact_value.nil?
-
-        core_value = Facter.core_value(name) if fact.used_resolution_weight <= 0
-        core_value.nil? ? fact_value : core_value
+        # fact_value = fact&.value
+        # return Facter.core_value(name) if fact_value.nil?
+        #
+        # core_value = Facter.core_value(name) if fact.used_resolution_weight <= 0
+        # core_value.nil? ? fact_value : core_value
```
Comment (IrimieBogdan) [inline lib/custom_facts/util/collection.rb:154]: The commented code should be recmoved

File: lib/custom_facts/util/fact.rb [hunk 2]
```diff
@@ -123,7 +123,7 @@ def value
           announce_when_no_suitable_resolution(suitable_resolutions)
           announce_when_no_value_found(@value)
 
-          @value
+          resolve_value
```
Comment (IrimieBogdan) [inline lib/custom_facts/util/fact.rb:126]: We should assign the resolve_value to `@value`. By doing this, the fact will only be resolved once, even if it is called multiple times.

File: lib/custom_facts/util/directory_loader.rb [hunk 3]
```diff
@@ -30,55 +30,91 @@ class NoSuchDirectoryError < RuntimeError
...
-          LegacyFacter::Util::DirectoryLoader.new(dir)
-        end
-        LegacyFacter::Util::CompositeLoader.new(loaders)
-      end
+      # def self.default_loader
+      #   loaders = LegacyFacter::Util::Config.external_facts_dirs.collect do |dir|
+      #     LegacyFacter::Util::DirectoryLoader.new(dir)
+      #   end
+      #   LegacyFacter::Util::CompositeLoader.new(loaders)
+      # end
```
Comment (IrimieBogdan) [inline lib/custom_facts/util/directory_loader.rb:51]: Commented code should be removed.

File: lib/custom_facts/util/directory_loader.rb [hunk 4]
```diff
@@ -30,55 +30,91 @@ class NoSuchDirectoryError < RuntimeError
...
+        cm = Facter::CacheManager.new
+        sf = []
         entries.each do |file|
-          parser = LegacyFacter::Util::Parser.parser_for(file)
+          basename = File.basename(file)
+          if sf.find { |f| f.name == basename } && cm.group_cached?(basename)
+            Facter.log_exception(Exception.new("Caching is enabled for group \"#{basename}\" while "\
+              'there are at least two external facts files with the same filename'))
+          else
+            f = Facter::SearchedFact.new(basename, nil, [], nil, :file)
```
Comment (IrimieBogdan) [inline lib/custom_facts/util/directory_loader.rb:79]: rename `f` to something more suggestive e.g. `searched_fact`

File: lib/custom_facts/util/directory_loader.rb [hunk 5]
```diff
@@ -30,55 +30,91 @@ class NoSuchDirectoryError < RuntimeError
...
       # parse them.
       def load(collection)
         weight = @weight
+
+        searched_facts, cached_facts = load_directory_entries(collection)
+
+        cached_facts.each do |cached_fact|
+          collection.add(cached_fact.name, value: cached_fact.value, fact_type: :external,
+                                           file: cached_fact.file) { has_weight(weight) }
+        end
```
Comment (IrimieBogdan) [inline lib/custom_facts/util/directory_loader.rb:63]: We could extract this in a method e.g.
```
      def load_cached_facts(collection, cached_facts, weight)
        cached_facts.each do |cached_fact|
          collection.add(cached_fact.name, value: cached_fact.value, fact_type: :external,
                         file: cached_fact.file) { has_weight(weight) }
        end
      end
```

File: lib/custom_facts/util/directory_loader.rb [hunk 6]
```diff
@@ -30,55 +30,91 @@ class NoSuchDirectoryError < RuntimeError
...
           end
         end
       end
 
-      private
-
       def entries
-        Dir.entries(directory).find_all { |f| should_parse?(f) }.sort.map { |f| File.join(directory, f) }
+        dirs = @directories.select { |directory| File.directory?(directory) }.map do |directory|
+          Dir.entries(directory).map { |d| directory + '/' + d }
```
Comment (IrimieBogdan) [inline lib/custom_facts/util/directory_loader.rb:100]: `d` should be renamed to something suggestive like `directory_entries`

File: external_facts/executable_external_fact.sh [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+#echo "SCRIPT STDERR" >&2
```
Comment (Filipovici-Andrei) [inline external_facts/executable_external_fact.sh:2]: probably this can be removed

File: lib/framework/core/cache_manager.rb [hunk 8]
```diff
@@ -13,14 +13,17 @@ def resolve_facts(searched_facts)
...
+          false
+        end
       end
-      facts.each do |fact|
-        searched_facts.delete_if { |f| f.name == fact.name }
-      end
-      [searched_facts, facts]
+
+      [searched_facts, facts.flatten]
     end
```
Comment (IrimieBogdan) [inline lib/framework/core/cache_manager.rb:27]: We could refactor this method e.g.

```
    def resolve_facts(searched_facts)
      return searched_facts, [] if !File.directory?(@cache_dir) || !Options[:cache]

      cached_facts = searched_facts.clone
      cached_facts.select! { |fact| resolve_fact(fact) }

      [searched_facts - cached_facts, cached_facts]
    end
```

File: lib/custom_facts/util/directory_loader.rb [hunk 9]
```diff
@@ -30,55 +30,82 @@ class NoSuchDirectoryError < RuntimeError
...
       end
 
-      def self.default_loader
-        loaders = LegacyFacter::Util::Config.external_facts_dirs.collect do |dir|
-          LegacyFacter::Util::DirectoryLoader.new(dir)
+      private
+
+      def load_directory_entries(_collection)
+        cm = Facter::CacheManager.new
+        sf = []
```
Comment (Filipovici-Andrei) [inline lib/custom_facts/util/directory_loader.rb:56]: probably facts is more suitable than sf

File: lib/custom_facts/util/directory_loader.rb (hunk 6 above)
Comment (Filipovici-Andrei) [reply lib/custom_facts/util/directory_loader.rb:100]: And a file join would be good.

File: lib/custom_facts/util/fact.rb [hunk 10]
```diff
@@ -123,8 +123,10 @@ def value
           announce_when_no_suitable_resolution(suitable_resolutions)
           announce_when_no_value_found(@value)
 
-          @value
+          @value = resolve_value
```
Comment (Filipovici-Andrei) [inline lib/custom_facts/util/fact.rb:126]: isn't this overwriting line 121?

File: lib/framework/core/cache_manager.rb [hunk 11]
```diff
@@ -13,14 +13,17 @@ def resolve_facts(searched_facts)
       return searched_facts, [] if !File.directory?(@cache_dir) || !Options[:cache]
 
       facts = []
-      searched_facts.each do |fact|
+      searched_facts.delete_if do |fact|
         res = resolve_fact(fact)
-        facts << res if res
+        if res
+          facts << res
+          true
```
Comment (Filipovici-Andrei) [inline lib/framework/core/cache_manager.rb:20]: are the booleans needed?

File: lib/framework/core/fact_loaders/external_fact_loader.rb [hunk 12]
```diff
@@ -34,8 +34,14 @@ def load_external_facts
 
       external_facts_to_load = LegacyFacter.collection.external_facts
 
-      external_facts_to_load&.each do |external_fact_name|
-        loaded_fact = LoadedFact.new(external_fact_name.to_s, nil, :external)
+      # external_facts_to_load&.each do |external_fact_name|
```
Comment (Filipovici-Andrei) [inline lib/framework/core/fact_loaders/external_fact_loader.rb:37]: commented code

File: lib/framework/parsers/query_parser.rb [hunk 13]
```diff
@@ -26,7 +26,9 @@ def parse(query_list, loaded_fact)
         matched_facts = []
         @log.debug "User query is: #{query_list}"
         @query_list = query_list
-        query_list = loaded_fact.map(&:name) unless query_list.any?
+        # query_list = loaded_fact.map(&:name) unless query_list.any?
```
Comment (Filipovici-Andrei) [inline lib/framework/parsers/query_parser.rb:29]: commented code

File: spec/custom_facts/util/directory_loader_spec.rb [hunk 14]
```diff
@@ -12,24 +12,24 @@
   let(:collection_double) { instance_spy(LegacyFacter::Util::Collection) }
 
   it 'makes the directory available' do
-    expect(dir_loader.directory).to be_instance_of(String)
+    expect(dir_loader.directories).to be_instance_of(Array)
   end
 
   # it "can be created with a given directory" do
```
Comment (Filipovici-Andrei) [inline spec/custom_facts/util/directory_loader_spec.rb:18]: commented test

File: spec/custom_facts/util/directory_loader_spec.rb [hunk 15]
```diff
@@ -12,24 +12,24 @@
...
 
-  it 'raises an error when the directory does not exist' do
-    missing_dir = 'missing'
-    allow(File).to receive(:directory?).with(missing_dir).and_return(false)
-    expect { LegacyFacter::Util::DirectoryLoader.loader_for(missing_dir) }
-      .to raise_error LegacyFacter::Util::DirectoryLoader::NoSuchDirectoryError
-  end
+  # it 'raises an error when the directory does not exist' do
+  #   missing_dir = 'missing'
+  #   allow(File).to receive(:directory?).with(missing_dir).and_return(false)
```
Comment (Filipovici-Andrei) [inline spec/custom_facts/util/directory_loader_spec.rb:24]: commented test

File: lib/framework/core/cache_manager.rb [hunk 16]
```diff
@@ -51,16 +65,32 @@ def resolve_fact(searched_fact)
       return unless data
 
       @log.debug("loading cached values for #{group_name} facts")
-      create_fact(searched_fact, data[searched_fact.name])
+
+      create_fact(searched_fact, data)
```
Comment (IrimieBogdan) [inline lib/framework/core/cache_manager.rb:69]: `create_facts`

File: lib/framework/core/cache_manager.rb [hunk 17]
```diff
@@ -51,16 +65,32 @@ def resolve_fact(searched_fact)
...
     end
 
-    def create_fact(searched_fact, value)
-      Facter::ResolvedFact.new(searched_fact.name, value, searched_fact.type,
-                               searched_fact.user_query, searched_fact.filter_tokens)
+    def create_fact(searched_fact, data)
+      if searched_fact.type == :file
+        facts = []
+        data.each do |k, v|
```
Comment (IrimieBogdan) [inline lib/framework/core/cache_manager.rb:75]: `k` => fact_name
`v` => fact_value

File: spec/facter/cache_manager_spec.rb [hunk 18]
```diff
@@ -6,11 +6,11 @@
   let(:cache_dir) { '/etc/facter/cache' }
```
Comment (IrimieBogdan) [inline spec/facter/cache_manager_spec.rb:6]: Add test for `group_cached?` The test for `resolve_facts` should be extended to cover external facts (when file is present)

File: facter.conf [hunk 19]
```diff
@@ -1,25 +1,6 @@
...
-
-cli : {
-  debug     : false,
-  trace     : true,
-  verbose   : false,
-  log-level : "warn"
-}
-
-fact-groups : {
-  cached-custom-facts : ["my_custom_fact"],
```
Comment (oanatmaria) [inline facter.conf:24]: these should be restored

File: lib/framework/core/cache_manager.rb (hunk 11 above)
Comment (florindragos) [reply lib/framework/core/cache_manager.rb:20]: facts are being deleted from searched_facts based on the return of the block

**Review by BogdanIrimie — APPROVED** [review]

**Pull request merged**